linux/vhba_hot_add_remove/get_new_disk_node_existing_ctrl.yml

```diff
@@ -3,50 +3,69 @@
   set_fact:
-    found_existing_ctrl_number: "{{ item.key }}"
-    device_key: "{{ item.value['controller_disks_devicekey'] }}"
-    found_existing_ctrl: True
-  with_dict: "{{ scsi_facts }}"
+    new_ctrl_number: "{{ new_disk_ctrl_facts[item].controller_busnumber }}"
+    found_existing_ctrl: "{{ new_disk_ctrl_facts[item] }}"
   when:
-    - found_existing_ctrl is defined and not found_existing_ctrl
-    - scsi_name_map[scsi_type] == item.value['controller_summary']
-- name: "Print the result of finding '{{ scsi_type }}' controller"
+    - controller_name_map[new_disk_ctrl_type] == new_disk_ctrl_facts[item].controller_summary
+    - new_disk_ctrl_facts[item].controller_disks_devicekey | length < 16
+  with_items: "{{ new_disk_ctrl_facts }}"
+
+- name: "Found an existing '{{ new_disk_ctrl_type }}' controller"
   debug: var=found_existing_ctrl
 
 - block:
-    - name: "Get total number of existing disk"
-      set_fact:
-        total_existing_disks: "{{ device_key | length }}"
-        new_unit_number: "0"
-        new_ctrl_number: "{{ found_existing_ctrl_number }}"
     - block:
-        - name: "Loop to get new disk number not assigned"
+        - include_tasks: ../../common/vm_get_disk_facts.yml
+
+        - name: "Get all disks attached to {{ found_existing_ctrl.controller_label }}"
+          set_fact:
+            existing_disks: "{{ vm_guest_disk_facts.guest_disk_info.values() | selectattr('controller_key', 'equalto', found_existing_ctrl.controller_devicekey) }}"
+
+        - name: "Get all occupied disk unit numbers"
+          set_fact:
+            existing_disk_unit_numbers: "{{ existing_disks | map(attribute='unit_number') | list }}"
+      when: new_disk_ctrl_type != 'sata'
+
+    - block:
+        - include_tasks: ../../common/vm_get_devices_with_ctrl_key.yml
+          vars:
+            device_controller_key: "{{ found_existing_ctrl.controller_devicekey }}"
+
+        - name: "Get all devices attached to {{ found_existing_ctrl.controller_label }}"
           set_fact:
-            new_unit_number: "{{ item }}"
-            found_new_disk: True
-          loop: "{{ range(0, 16) | list }}"
-          when:
-            - total_existing_disks | int >= 1
-            - "(item + 16 * new_ctrl_number|int + 2000) not in device_key"
-            - not found_new_disk
-        - debug:
-            msg:
-              - "Set new_ctrl_number: {{ new_ctrl_number }}"
-              - "Set new_unit_number: {{ new_unit_number }}"
-      when: total_existing_disks | int < 16
+            existing_disks: "{{ device_list_with_ctrl_key }}"
+
+        - name: "Get all occupied disk unit numbers"
+          set_fact:
+            existing_disk_unit_numbers: "{{ existing_disks | map(attribute='unitNumber') | list }}"
+      when: new_disk_ctrl_type == 'sata'
+
+    - name: "Set the new disk controller unit number"
+      set_fact:
+        new_unit_number: "{{ range(0, 16) | difference(existing_disk_unit_numbers) | min }}"
```

Comment:
For SATA and NVMe controller, it has different disk unit limitation other than 16.

Reply:
We have a common task `common/vm_get_new_disk_node_info.yml ` to do this.

Reply:
We will only cover new disk on new controller, so the new unit number will be 0.